salesagility/SuiteCRM
# Fix #10629 - Duplicating Surveys

Closes https://github.com/salesagility/SuiteCRM/issues/8947

As the Issue mentions, there is a problem when duplicating Surveys. Basically, what goes wrong when duplicating:
- If the base survey has questions and no answers, the duplicated survey will take all the questions, leaving the base survey without questions
- If the base survey has questions and answers, the duplicated survey won't take/keep the questions, neither the answers. Although the Editview that appears after duplicating will contain the questions and won't allow it to modify them:
![Selection_430](https://user-images.githubusercontent.com/61022311/152755020-27103130-3d28-4c47-8959-217fd51c1e73.png)


Closes https://github.com/salesagility/SuiteCRM/issues/10629

Having a WorkFlow and duplicating a survey or create also duplicated the questions, as the code was executed without any workflow verification. This was because:
- When duplicating survey X to Y or creating survey X, having WorkFlow active, the code saved the questions normally.
- But because there was no control over the workflow, when the workflow was executed:
   - The code re-executed the foreach loop that saves the questions.
   - As a result, the questions were saved twice: once during the normal save and once during the workflow execution.

## Description
This PR fixes the Duplicate functionality for the Surveys module, as we understand it should work:
- If the base survey has questions and no answers, the duplicated survey will keep the questions, duplicating them.
- If the base survey has questions and answers, the duplicated survey ~won't~ will take/keep/duplicate the questions ~or~ but not the answers. And other new Questions can be added during the first edition after duplication.
- Fixed the workflow problem by adding a control variable `already_saved` which is used to verify that the method is not re-executed. If `already_saved` is set, it means that the method has already been executed and reprocessing of the questions is avoided.

## Motivation and Context
With this change, duplicating Surveys will be consistent.

## How To Test This
1- Create a Survey
2- Add questions
3- Duplicate it.
4- Repeat adding answers to the survey.

- Workflow Issue:
1. WorkFlow Module: Surveys
2. Run: Only On Save
3. Run On: New Records
4. Action:
  4.1. Modify the survey itself, for example, change from draft to published status.
  4.2. Calculate fields, for example, calculate the name.
5. Duplicate the survey from step 1, checking the number of questions it has.
6. Check that the survey has been duplicated without duplicating the questions and that it has the same questions as the survey from step 1.
7. Create a survey with several questions and check that the surveys are not duplicated.

## Types of changes

### Final checklist
